File: examples/hello-world/gabriel-greeting-app-flutter/app/lib/src/controller/coax_controller.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:grpc/grpc.dart';
import 'package:holons/holons.dart' as holons;

import '../model/app_model.dart';
import '../rpc/coax_service.dart';
import '../rpc/greeting_app_service.dart';
import '../runtime/describe_registration.dart';
import '../settings_store.dart';
import 'greeting_controller.dart';

class CoaxController extends ChangeNotifier {
  CoaxController({
    required GreetingController greetingController,
    required SettingsStore settingsStore,
    AppPlatformCapabilities? capabilities,
  }) : _greetingController = greetingController,
       _settingsStore = settingsStore,
       capabilities = capabilities ?? AppPlatformCapabilities.desktopCurrent(),
       _isEnabled = settingsStore.readBool(_enabledKey),
       _snapshot = _loadSnapshot(settingsStore) {
    if (!this.capabilities.supportsUnixSockets &&
        _snapshot.serverTransport == CoaxServerTransport.unix) {
      _snapshot = CoaxSettingsSnapshot.defaults;
    }
  }

  static const _enabledKey = 'coax.server.enabled';
  static const _settingsKey = 'coax.server.settings';

  final GreetingController _greetingController;
  final SettingsStore _settingsStore;
  final AppPlatformCapabilities capabilities;

  holons.RunningServer? _runningServer;
  bool _disposed = false;
  bool _isEnabled;
  CoaxSettingsSnapshot _snapshot;
  String? listenUri;
  String? statusDetail;
  int _startGeneration = 0;

  bool get isEnabled => _isEnabled;
  bool get serverEnabled => _snapshot.serverEnabled;
  CoaxServerTransport get serverTransport => _snapshot.serverTransport;
  String get serverHost => _snapshot.serverHost;
  String get serverPortText => _snapshot.serverPortText;
  String get serverUnixPath => _snapshot.serverUnixPath;

  int get serverPort =>
      sanitizedPort(serverPortText, serverTransport.defaultPort);

  String? get serverPortValidationMessage {
    if (serverTransport != CoaxServerTransport.tcp) {
      return null;
    }
    final trimmed = serverPortText.trim();
    if (trimmed.isEmpty) {
      return 'Empty port. Falling back to ${serverTransport.defaultPort}.';
    }
    final parsed = int.tryParse(trimmed);
    if (parsed == null || parsed < 1 || parsed > 65535) {
      return 'Invalid port. Falling back to ${serverTransport.defaultPort}.';
    }
    return null;
  }

  String get serverPreviewEndpoint {
    switch (serverTransport) {
      case CoaxServerTransport.tcp:
        final host = serverHost.trim().isEmpty
            ? CoaxSettingsSnapshot.defaultHost
            : serverHost.trim();
        return 'tcp://$host:$serverPort';
      case CoaxServerTransport.unix:
        final path = serverUnixPath.trim().isEmpty
            ? CoaxSettingsSnapshot.defaultUnixPath
            : serverUnixPath.trim();
        return 'unix://$path';
    }
  }

  CoaxSurfaceStatus get serverStatus {
    return CoaxSurfaceStatus(
      id: 'server',
      title: 'Server',
      endpoint: serverEnabled
          ? (listenUri ?? serverPreviewEndpoint)
          : serverPreviewEndpoint,
      state: _serverSurfaceState,
    );
  }

  CoaxSurfaceState get _serverSurfaceState {
    if (!serverEnabled) {
      return CoaxSurfaceState.off;
    }
    if (statusDetail != null && isEnabled) {
      return CoaxSurfaceState.error;
    }
    if (listenUri != null) {
      return CoaxSurfaceState.live;
    }
    if (!isEnabled) {
      return CoaxSurfaceState.saved;
    }
    return CoaxSurfaceState.announced;
  }

  Future<void> startIfEnabled() async {
    if (!_isEnabled) {
      return;
    }
    await _reconfigureRuntime();
  }

  Future<void> setIsEnabled(bool value) async {
    if (_isEnabled == value) {
      return;
    }
    _isEnabled = value;
    await _settingsStore.writeBool(_enabledKey, value);
    _safeNotify();
    await _reconfigureRuntime();
  }

  Future<void> setServerEnabled(bool value) async {
    if (serverEnabled == value) {
      return;
    }
    _snapshot = CoaxSettingsSnapshot(
      serverEnabled: value,
      serverTransport: serverTransport,
      serverHost: serverHost,
      serverPortText: serverPortText,
      serverUnixPath: serverUnixPath,
    );
    await _persistSnapshot();
    _safeNotify();
    await _reconfigureRuntime();
  }

  Future<void> setServerTransport(CoaxServerTransport value) async {
    if (value == serverTransport) {
      return;
    }
    final nextPortText = _usesDefaultServerPort(serverTransport)
        ? value.defaultPort.toString()
        : serverPortText;
    _snapshot = CoaxSettingsSnapshot(
      serverEnabled: serverEnabled,
      serverTransport: value,
      serverHost: serverHost,
      serverPortText: nextPortText,
      serverUnixPath: serverUnixPath,
    );
    await _persistSnapshot();
    _safeNotify();
    if (serverEnabled) {
      await _reconfigureRuntime();
    }
  }

  Future<void> setServerHost(String value) async {
    if (value == serverHost) {
      return;
    }
    _snapshot = CoaxSettingsSnapshot(
      serverEnabled: serverEnabled,
      serverTransport: serverTransport,
      serverHost: value,
      serverPortText: serverPortText,
      serverUnixPath: serverUnixPath,
    );
    await _persistSnapshot();
    _safeNotify();
    if (serverEnabled) {
      await _reconfigureRuntime();
    }
  }

  Future<void> setServerPortText(String value) async {
    if (value == serverPortText) {
      return;
    }
    _snapshot = CoaxSettingsSnapshot(
      serverEnabled: serverEnabled,
      serverTransport: serverTransport,
      serverHost: serverHost,
      serverPortText: value,
      serverUnixPath: serverUnixPath,
    );
    await _persistSnapshot();
    _safeNotify();
    if (serverEnabled) {
      await _reconfigureRuntime();
    }
  }

  Future<void> setServerUnixPath(String value) async {
    if (value == serverUnixPath) {
      return;
    }
    _snapshot = CoaxSettingsSnapshot(
      serverEnabled: serverEnabled,
      serverTransport: serverTransport,
      serverHost: serverHost,
      serverPortText: serverPortText,
      serverUnixPath: value,
    );
    await _persistSnapshot();
    _safeNotify();
    if (serverEnabled) {
      await _reconfigureRuntime();
    }
  }

  Future<void> disableAfterRpc() async {
    await Future<void>.delayed(const Duration(milliseconds: 100));
    await setIsEnabled(false);
  }

  Future<void> shutdown() async {
    _disposed = true;
    await _stopServer(clearStatus: true);
  }

  Future<void> _reconfigureRuntime() async {
    _startGeneration += 1;
    final generation = _startGeneration;

    if (!_isEnabled || !serverEnabled) {
      await _stopServer(clearStatus: true);
      return;
    }

    await _stopServer(clearStatus: false);
    await _startServer(generation);
  }

  Future<void> _startServer(int generation) async {
    final appService = GreetingAppRpcService(_greetingController);
    final coaxService = CoaxRpcService(
      greetingController: _greetingController,
      coaxController: this,
    );
    listenUri = null;
    statusDetail = null;
    _safeNotify();

    try {
      ensureAppDescribeRegistered();
      final server = await holons.startWithOptions(
        _runtimeListenUri(),
        <Service>[coaxService, appService],
        options: const holons.ServeOptions(describe: true, logger: _ignoreLog),
      );
      if (generation != _startGeneration || !_isEnabled || !serverEnabled) {
        await server.stop();
        return;
      }
      _runningServer = server;
      listenUri = server.publicUri;
      statusDetail = null;
      _log('[COAX] server listening on ${server.publicUri}');
    } on Object catch (error) {
      if (generation != _startGeneration) {
        return;
      }
      listenUri = null;
      statusDetail = 'Server surface failed to start: $error';
      _log('[COAX] failed to start server: $error');
    }

    _safeNotify();
  }

  Future<void> _stopServer({required bool clearStatus}) async {
    final server = _runningServer;
    _runningServer = null;
    listenUri = null;
    if (clearStatus) {
      statusDetail = null;
    }
    _safeNotify();
    if (server == null) {
      return;
    }
    _log('[COAX] server stopped');
    await Future<void>.delayed(const Duration(milliseconds: 250));
    await server.stop();
  }

  Future<void> _persistSnapshot() async {
    await _settingsStore.writeString(_settingsKey, _snapshot.encode());
  }

  String _runtimeListenUri() {
    switch (serverTransport) {
      case CoaxServerTransport.tcp:
        final host = serverHost.trim().isEmpty
            ? CoaxSettingsSnapshot.defaultHost
            : serverHost.trim();
        return 'tcp://$host:$serverPort';
      case CoaxServerTransport.unix:
        final path = serverUnixPath.trim().isEmpty
            ? CoaxSettingsSnapshot.defaultUnixPath
            : serverUnixPath.trim();
        return 'unix://$path';
    }
  }

  bool _usesDefaultServerPort(CoaxServerTransport transport) {
    if (transport == CoaxServerTransport.unix) {
      return serverPortText.trim().isEmpty;
    }
    final trimmed = serverPortText.trim();
    if (trimmed.isEmpty) {
      return true;
    }
    return sanitizedPort(trimmed, transport.defaultPort) ==
        transport.defaultPort;
  }

  static CoaxSettingsSnapshot _loadSnapshot(SettingsStore store) {
    final rawValue = store.readString(_settingsKey);
    if (rawValue.trim().isEmpty) {
      return CoaxSettingsSnapshot.defaults;
    }
    try {
      return CoaxSettingsSnapshot.decode(rawValue);
    } on Object {
      return CoaxSettingsSnapshot.defaults;
    }
  }

  void _safeNotify() {
    if (!_disposed) {
      notifyListeners();
    }
  }

  void _log(String message) {
    stderr.writeln(message);
  }

  static void _ignoreLog(String _) {}

  @override
  void dispose() {
    _disposed = true;
    super.dispose();
  }
}

File: examples/hello-world/gabriel-greeting-app-kotlin-compose/app/src/main/kotlin/org/organicprogramming/gabriel/greeting/kotlincompose/controller/CoaxController.kt
package org.organicprogramming.gabriel.greeting.kotlincompose.controller

import io.grpc.BindableService
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.organicprogramming.gabriel.greeting.kotlincompose.model.AppPlatformCapabilities
import org.organicprogramming.gabriel.greeting.kotlincompose.model.CoaxServerTransport
import org.organicprogramming.gabriel.greeting.kotlincompose.model.CoaxSettingsSnapshot
import org.organicprogramming.gabriel.greeting.kotlincompose.model.CoaxUiState
import org.organicprogramming.gabriel.greeting.kotlincompose.runtime.DescribeRegistration
import org.organicprogramming.gabriel.greeting.kotlincompose.rpc.CoaxRpcService
import org.organicprogramming.gabriel.greeting.kotlincompose.rpc.GreetingAppRpcService
import org.organicprogramming.gabriel.greeting.kotlincompose.settings.SettingsStore
import org.organicprogramming.gabriel.greeting.kotlincompose.settings.coaxEnabledKey
import org.organicprogramming.gabriel.greeting.kotlincompose.settings.coaxSettingsKey
import org.organicprogramming.holons.Serve

class CoaxController(
    private val greetingController: GreetingController,
    private val settingsStore: SettingsStore,
    private val capabilities: AppPlatformCapabilities = AppPlatformCapabilities.desktopCurrent(),
    dispatcher: CoroutineDispatcher = Dispatchers.IO,
) {
    private val scope = CoroutineScope(SupervisorJob() + dispatcher)
    private val _state = MutableStateFlow(
        CoaxUiState(
            isEnabled = settingsStore.readBool(coaxEnabledKey()),
            snapshot = loadSnapshot(settingsStore, capabilities),
            capabilities = capabilities,
        ),
    )

    val state: StateFlow<CoaxUiState> = _state.asStateFlow()

    private var runningServer: Serve.RunningServer? = null
    private var disposed = false
    private var startGeneration = 0

    suspend fun startIfEnabled() {
        if (!state.value.isEnabled) {
            return
        }
        reconfigureRuntime()
    }

    suspend fun setIsEnabled(value: Boolean) {
        if (state.value.isEnabled == value) {
            return
        }
        settingsStore.writeBool(coaxEnabledKey(), value)
        _state.update { it.copy(isEnabled = value) }
        reconfigureRuntime()
    }

    suspend fun setServerEnabled(value: Boolean) {
        if (state.value.serverEnabled == value) {
            return
        }
        updateSnapshot(state.value.snapshot.copy(serverEnabled = value))
        reconfigureRuntime()
    }

    suspend fun setServerTransport(value: CoaxServerTransport) {
        if (state.value.serverTransport == value) {
            return
        }
        val nextPortText = if (usesDefaultServerPort()) value.defaultPort.toString() else state.value.serverPortText
        updateSnapshot(
            state.value.snapshot.copy(
                serverTransport = value,
                serverPortText = nextPortText,
            ),
        )
        if (state.value.serverEnabled) {
            reconfigureRuntime()
        }
    }

    suspend fun setServerHost(value: String) {
        if (state.value.serverHost == value) {
            return
        }
        updateSnapshot(state.value.snapshot.copy(serverHost = value))
        if (state.value.serverEnabled) {
            reconfigureRuntime()
        }
    }

    suspend fun setServerPortText(value: String) {
        if (state.value.serverPortText == value) {
            return
        }
        updateSnapshot(state.value.snapshot.copy(serverPortText = value))
        if (state.value.serverEnabled) {
            reconfigureRuntime()
        }
    }

    suspend fun setServerUnixPath(value: String) {
        if (state.value.serverUnixPath == value) {
            return
        }
        updateSnapshot(state.value.snapshot.copy(serverUnixPath = value))
        if (state.value.serverEnabled) {
            reconfigureRuntime()
        }
    }

    fun disableAfterRpc() {
        scope.launch {
            delay(100)
            setIsEnabled(false)
        }
    }

    suspend fun shutdown() {
        disposed = true
        stopServer(clearStatus = true)
        scope.cancel()
    }

    private suspend fun reconfigureRuntime() {
        val generation = ++startGeneration
        if (!state.value.isEnabled || !state.value.serverEnabled) {
            stopServer(clearStatus = true)
            return
        }

        stopServer(clearStatus = false)
        startServer(generation)
    }

    private suspend fun startServer(generation: Int) {
        val appService: BindableService = GreetingAppRpcService(greetingController)
        val coaxService: BindableService = CoaxRpcService(greetingController, this)
        _state.update { it.copy(listenUri = null, statusDetail = null) }

        try {
            DescribeRegistration.ensureRegistered()
            val server = Serve.startWithOptions(
                runtimeListenUri(),
                listOf(coaxService, appService),
                Serve.Options(describe = true, logger = { }),
            )
            if (generation != startGeneration || !state.value.isEnabled || !state.value.serverEnabled) {
                server.stop()
                return
            }
            runningServer = server
            _state.update { it.copy(listenUri = server.publicUri, statusDetail = null) }
        } catch (error: Throwable) {
            if (generation != startGeneration) {
                return
            }
            _state.update {
                it.copy(
                    listenUri = null,
                    statusDetail = "Server surface failed to start: $error",
                )
            }
        }
    }

    private suspend fun stopServer(clearStatus: Boolean) {
        val server = runningServer
        runningServer = null
        _state.update {
            it.copy(
                listenUri = null,
                statusDetail = if (clearStatus) null else it.statusDetail,
            )
        }
        if (server == null) {
            return
        }
        delay(250)
        server.stop()
    }

    private fun runtimeListenUri(): String =
        when (state.value.serverTransport) {
            CoaxServerTransport.TCP -> {
                val host = state.value.serverHost.trim().ifEmpty { CoaxSettingsSnapshot.defaultHost }
                "tcp://$host:${state.value.serverPort}"
            }
            CoaxServerTransport.UNIX -> {
                val path = state.value.serverUnixPath.trim().ifEmpty { CoaxSettingsSnapshot.defaultUnixPath }
                "unix://$path"
            }
        }

    private fun usesDefaultServerPort(): Boolean {
        val trimmed = state.value.serverPortText.trim()
        if (state.value.serverTransport == CoaxServerTransport.UNIX) {
            return trimmed.isEmpty()
        }
        return trimmed.isEmpty() || trimmed.toIntOrNull() == state.value.serverTransport.defaultPort
    }

    private fun updateSnapshot(snapshot: CoaxSettingsSnapshot) {
        settingsStore.writeString(coaxSettingsKey(), snapshot.encode())
        _state.update { it.copy(snapshot = snapshot) }
    }

    companion object {
        private fun loadSnapshot(
            store: SettingsStore,
            capabilities: AppPlatformCapabilities,
        ): CoaxSettingsSnapshot {
            val rawValue = store.readString(coaxSettingsKey())
            val decoded = CoaxSettingsSnapshot.decode(rawValue)
            return if (!capabilities.supportsUnixSockets && decoded.serverTransport == CoaxServerTransport.UNIX) {
                CoaxSettingsSnapshot.defaults
            } else {
                decoded
            }
        }
    }
}
